### ISSUE TITLE
Building documentation requires docker / manual IPython config

### ISSUE THREAD

**Issue by sdrave** [pymor/pymor#1000]
See https://github.com/pymor/pymor/pull/984#issuecomment-654311513 .

Maybe the config can be automatically injected when `conf.py` is loaded?

**Comment by renefritze** [comment]
TBH I think it's not practical to ensure docs are always buildable w/o the controlled docker environment. With more tutorials incoming that may need external solvers, or pytorch, this will only get more complex. You would also need to add a CI job that tests this again, which doesn't depend on the docker infrastructure while also having all the deps present. In the light of active branches always having documentation online now, I feel this would be a lot of effort for very little gain. One can use rootless docker or podman too if running our images with super user privs is a concern. 
OTOH something like xfailing tutorials might be something you can look into, if using some container tech is strictly out of the question for yourself.

**Comment by sdrave** [comment]
You are basically saying that pyMOR development, or at least writing documentation for pyMOR requires using docker/podman. If find that very hard to accept and not think it is necessary.

> With more tutorials incoming that may need external solvers, or pytorch, this will only get more complex.

So far I never had trouble installing all external solvers supported by pyMOR. Also, pytorch is very easy to install. In any case, I do not necessarily want to be able to build all the documentation there is, but at least I want to be able to build the documentation I am currently working on and for that all dependencies will be fulfilled.

> You would also need to add a CI job that tests this again, which doesn't depend on the docker infrastructure while also having all the deps present.

I do not see the need for having CI that checks if docs can be built in a non-docker setup. Building the docs is mainly of interest for developers and those can live with occasional doc building issues.

> One can use rootless docker or podman too if running our images with super user privs is a concern.

Not sure how 'rootless docker' works, but podman packages for debian are still work in progress. But this is not my main concern. Apart from disliking container-based workflows for software development: it is one thing that we rely on containers to make CI/CD manageable/reproducible. But we are not developing an application but a library that should work in many different development environment. So if something does not work outside the docker image, the answer should not be 'use docker' but we should try to solve the issue.

> OTOH something like xfailing tutorials might be something you can look into, if using some container tech is strictly out of the question for yourself.

The issue here has nothing to do with individual tutorials that are failing but with a broken build setup unless some arcane IPython config is applied. But yes, when tutorials are hooked up into pytest, then tutorials requiring external solvers should xfail like current tests are already doing.

**Comment by renefritze** [comment]
> You are basically saying that pyMOR development, or at least writing documentation for pyMOR requires using docker/podman. 

I very clearly did not say that. I said not practical and even mentioned an alternative in allowing failures for tutorial builds.

> 
> > With more tutorials incoming that may need external solvers, or pytorch, this will only get more complex.
> 
> So far I never had trouble installing all external solvers supported by pyMOR. 

Is that a rebuttal against increasing complexity? Remember not everybody has your level of experience. 

> Also, pytorch is very easy to install. 

Only after you read additional info on which version you need, which either we have to provide or the user must find on her own. AFAICT we cannot [supply/control this](https://github.com/pymor/pymor/pull/1001#issuecomment-654704109) just with our python buildsystem. Same goes for most of the external solvers. 
My point here is to make everybody aware that we simply cannot control/check all aspects of the envirornment with means of pyMOR's build system. So in _general_ (not particularly with the output logger box) the solution might just be to have the user change the environment itself. 
For the particular issue you actually opened, the case is _not_ that the build needs docker. The issue is that the logger redirect needs to be loaded. The container environment just ensures that. 

> In any case, I do not necessarily want to be able to build all the documentation there is, but at least I want to be able to build the documentation I am currently working on and for that all dependencies will be fulfilled.

Currently not all dependencies for you are fulfilled though. That the extension is loaded is a dependency. A good question would be if we can change that, (We probably can) or how the loading can be automatic without adding the kernel config. 

> 
> > You would also need to add a CI job that tests this again, which doesn't depend on the docker infrastructure while also having all the deps present.
> 
> I do not see the need for having CI that checks if docs can be built in a non-docker setup. Building the docs is mainly of interest for developers and those can live with occasional doc building issues.

If you can live with this being generally broken, ok.

> 
> > One can use rootless docker or podman too if running our images with super user privs is a concern.
> 
> Not sure how 'rootless docker' works

Runs the docker daemon as your user. 

> But this is not my main concern. Apart from disliking container-based workflows for software development: it is one thing that we rely on containers to make CI/CD manageable/reproducible. But we are not developing an application but a library that should work in many different development environment. So if something does not work outside the docker image, the answer should not be 'use docker' but we should try to solve the issue.

I feel like you've constructed this conclusion for yourself. What I wanted to be read as could be summarized as "if you don't want to change your environment, use the docker one".

> 
> > OTOH something like xfailing tutorials might be something you can look into, if using some container tech is strictly out of the question for yourself.
> 
> The issue here has nothing to do with individual tutorials that are failing but with a broken build setup unless some arcane IPython config is applied. 

If you know a better way to activate kernel extensions that would be great! Or if you've found out why the logger output without it results in sphinx warnings with jupyter-sphinx.

**Comment by sdrave** [comment]
> > You are basically saying that pyMOR development, or at least writing documentation for pyMOR requires using docker/podman.
> 
> I very clearly did not say that. I said not practical and even mentioned an alternative in allowing failures for tutorial builds.

Sorry, but I understood your comment 

> TBH I think it's not practical to ensure docs are always buildable w/o the controlled docker environment.

to this issue as a suggestion to not try to fix the jupyter-spinx issue since supporting non-docker environments is not feasible / not worth the effort. This led to my above interpretation of the statement, which IMHO, would be a logical consequence of this position.

> > > With more tutorials incoming that may need external solvers, or pytorch, this will only get more complex.
> > 
> > 
> > So far I never had trouble installing all external solvers supported by pyMOR.
> 
> Is that a rebuttal against increasing complexity? Remember not everybody has your level of experience.

I am just saying that the picture is maybe not as dark as you draw it. Every external solver that is currently supported by pyMOR is relatively easy to install. I would expect every *developer*, maybe with some hints, to be able to do it. Of course, this can be tedious work, and it is great to have docker images available, where everything is already properly installed.

> > Also, pytorch is very easy to install.
> 
> Only after you read additional info on which version you need, which either we have to provide or the user must find on her own. 

I was only referring to developers. The instructions on the pytorch homepage are very easy to follow.

> AFAICT we cannot [supply/control this](https://github.com/pymor/pymor/pull/1001#issuecomment-654704109) just with our python buildsystem. Same goes for most of the external solvers.

I would never assume it to be the job of our buildsystem to install all external solvers.

> My point here is to make everybody aware that we simply cannot control/check all aspects of the envirornment with means of pyMOR's build system.
> So in _general_ (not particularly with the output logger box) the solution might just be to have the user change the environment itself.

For reproducible CI and so on, you need well-defined environments. However, pyMOR is a software library meant to be combined with other software in unknown ways. So, in my opinion, pointing the user to docker is not a good approach in general, except maybe for getting started or for educational purposes. And for that reason pyMOR should also be developed under as many different environments as possible, and if there is a problem with one environment, we should try to fix the problem.
 
> For the particular issue you actually opened, the case is _not_ that the build needs docker. The issue is that the logger redirect needs to be loaded. The container environment just ensures that.

Yes, I never claimed that the build needs docker (see the title of the issue). But the issue is that it is nowhere documented how to build the docs and that it is not nice to have to modify your IPython config to make it work.

> > In any case, I do not necessarily want to be able to build all the documentation there is, but at least I want to be able to build the documentation I am currently working on and for that all dependencies will be fulfilled.
> 
> Currently not all dependencies for you are fulfilled though. That the extension is loaded is a dependency. 

I was referring to external solvers I might not have installed.

> A good question would be if we can change that, (We probably can) or how the loading can be automatic without adding the kernel config.

And I already suggested to try to do something in `conf.py`. Another option would be to load the extension at pyMOR import.


> > > You would also need to add a CI job that tests this again, which doesn't depend on the docker infrastructure while also having all the deps present.
> > 
> > 
> > I do not see the need for having CI that checks if docs can be built in a non-docker setup. Building the docs is mainly of interest for developers and those can live with occasional doc building issues.
> 
> If you can live with this being generally broken, ok.

You mean 'generally working'. So far I rarely had relevant issues. 

> > But this is not my main concern. Apart from disliking container-based workflows for software development: it is one thing that we rely on containers to make CI/CD manageable/reproducible. But we are not developing an application but a library that should work in many different development environment. So if something does not work outside the docker image, the answer should not be 'use docker' but we should try to solve the issue.
> 
> I feel like you've constructed this conclusion for yourself. What I wanted to be read as could be summarized as "if you don't want to change your environment, use the docker one".

If there is a good reason why my environment does not work, that would be a good advice.

> > > OTOH something like xfailing tutorials might be something you can look into, if using some container tech is strictly out of the question for yourself.
> > 
> > 
> > The issue here has nothing to do with individual tutorials that are failing but with a broken build setup unless some arcane IPython config is applied.
> 
> If you know a better way to activate kernel extensions that would be great! Or if you've found out why the logger output without it results in sphinx warnings with jupyter-sphinx.

This is what I wanted to discuss in this issue. Now, I know nothing about jupyter-sphinx, but judging from

```
WARNING: Cell printed to stderr:
00:00 |WARNING|BitmapFunction: Image R.png not in grayscale mode. Converting to grayscale.
```

I would assume that jupyter-sphinx generally does not capture stderr output. I could manually activate the extension in the notebook by calling `load_ipython_extension` (at any point in execution time). So probably, it would be enough to call it when pyMOR is loaded in case some environment variable is set.

**Comment by renefritze** [comment]
> to this issue as a suggestion to not try to fix the jupyter-spinx issue since supporting non-docker environments is not feasible / not worth the effort. This led to my above interpretation of the statement, which IMHO, would be a logical consequence of this position.

The "ensuring" here is important, since that means automated testing. Automated testing for "most"  environments simply is not feasible. 


> I am just saying that the picture is maybe not as dark as you draw it. Every external solver that is currently supported by pyMOR is relatively easy to install.  I would expect every _developer_, maybe with some hints, to be able to do it. Of course, this can be tedious work, and it is great to have docker images available, where everything is already properly installed.

Just to check, for me something "relatively easy" that I expect any new developer can do with some hints is installing pre-made packages for her OS, possibly from a 3rd party source. 
Installing the petsc + fenics stack from source is qualitatively beyond that and I would not expect a new (python!) developer to able to do that. 

> > AFAICT we cannot [supply/control this](https://github.com/pymor/pymor/pull/1001#issuecomment-654704109) just with our python buildsystem. Same goes for most of the external solvers.
> 
> I would never assume it to be the job of our buildsystem to install all external solvers.
> 
Good. The point was more towards it's not even really possible to select the right pytorch to install.

> > My point here is to make everybody aware that we simply cannot control/check all aspects of the envirornment with means of pyMOR's build system.
> > So in _general_ (not particularly with the output logger box) the solution might just be to have the user change the environment itself.
> 
> For reproducible CI and so on, you need well-defined environments. However, pyMOR is a software library meant to be combined with other software in unknown ways. So, in my opinion, pointing the user to docker is not a good approach in general, except maybe for getting started or for educational purposes. And for that reason pyMOR should also be developed under as many different environments as possible, and if there is a problem with one environment, we should try to fix the problem.

I really shouldn't have said "user" here before since that's a somewhat different story. And I think we've agreed before that users don't build documentation. That aside, if a developer or user doesn't need more of a toolchain than we can offer in docker images, then using them is a great idea since they directly benefit from a stable environment that's continually tested. I don't think this should be limited to getting started or reference material, but should be evaluated with personal needs and wants in mind.

> > If you can live with this being generally broken, ok.
> 
> You mean 'generally working'. So far I rarely had relevant issues.

Alright. Only broken sometimes then :wink:

**Comment by sdrave** [comment]
Fixed in #1016.
